filegenerator/file_generator.go

```diff
@@ -0,0 +1,368 @@
+}
+
+func isValidDataDir(dataDirName string) bool {
+	matched, _ := regexp.MatchString("[0-9]{2}-[0-9]{2}-[0-9]{3}", dataDirName)
+	return matched
+}
+
+// RenderHistoricalBenchmarkResults takes the folders in benchSamples, parses them, and returns a
+// data structure called BenchPackages which represent the data from the folders.
+func RenderHistoricalBenchmarkResults(dirs []string) BenchPackages {
+	dirToTestNames := make(BenchPackages)
+	benchSamples := os.Getenv(benchSamplesEnv)
+	dataDirs, err := ioutil.ReadDir(benchSamples)
+	if err != nil {
+		log.Fatal(err)
+	}
+	for _, dataDir := range dataDirs {
+		if !dataDir.IsDir() {
+			continue
+		}
+		dataDirName := dataDir.Name()
+		if !isValidDataDir(dataDirName) {
+			continue
+		}
+		for _, dirName := range dirs {
+			if _, ok := dirToTestNames[dirName]; !ok {
+				dirToTestNames[dirName] = make(BenchTestMap)
+			}
+			fullDirPath := filepath.Join(benchSamples, dataDirName, "cockroach", dirName)
+			files, err := ioutil.ReadDir(fullDirPath)
+			if err != nil {
+				log.Fatal(err)
+			}
+			for _, file := range files {
+				fileName := file.Name()
+				matched, _ := regexp.MatchString(`.*test\.stdout`, fileName)
+				if matched {
+					content, err := ioutil.ReadFile(fullDirPath + "/" + fileName)
+					if err != nil {
+						log.Fatal(err)
+					}
+					for _, line := range strings.Split(string(content), "\n") {
+						if isValidTestLine(line) {
+							testName, stats := parseTestResultLine(line)
+							if _, ok := dirToTestNames[dirName][testName]; !ok {
+								dirToTestNames[dirName][testName] = make(BenchResults)
+							}
+							dirToTestNames[dirName][testName][dataDirName] = stats
+						}
+					}
+				}
+			}
+		}
+	}
+	return dirToTestNames
+}
+
+// GenerateJSONFiles takes a BenchPackages and creates a json file for every test
+// containing its benchmark results data over time.
+func GenerateJSONFiles(packages BenchPackages) {
+	deployRoot := os.Getenv(deployRootEnv)
+	for dir := range packages {
+		for testName := range packages[dir] {
+			testData := packages[dir][testName]
+			fileFullName := deployRoot + "/" + dir + "/" + testName + ".json"
+			testDataJSON, err := json.Marshal(&testData)
+			if err != nil {
+				log.Fatal(err)
+			}
+			err = ioutil.WriteFile(fileFullName, testDataJSON, 0644)
+			if err != nil {
+				log.Fatal(err)
+			}
+		}
+	}
+}
+
+// GenerateTestNameJSONFile creates a JSON file that represents a map of
+// directories to a list of all of the benchmark tests in that directory.
+func GenerateTestNameJSONFile(packages BenchPackages) {
+	deployRoot := os.Getenv(deployRootEnv)
+	testPackageToName := make(map[string][]string)
+	for packageName := range packages {
+		testPackageToName[packageName] = make([]string,
+			len(testPackageToName[packageName]))
+		for testName := range packages[packageName] {
+			testPackageToName[packageName] = append(testPackageToName[packageName], testName)
+		}
+	}
+	fileName := deployRoot + "/test_names.json"
```

Comment:
It's good to use `filepath.Join` whenever you're joining various path components. Among other things, it adds path separators only when necessary, which avoids some annoying possible errors.